src/prefect/engine.py

```diff
@@ -1418,7 +1412,7 @@ def resolve_input(expr):
         # Do not allow uncompleted upstreams except failures when `allow_failure` has
         # been used
         if not state.is_completed() and not (
-            should_allow_failure and state.is_failed()
+            isinstance(context.get("annotation"), allow_failure) and state.is_failed()
```

Comment:
Hm. This will break when annotations are nested e.g. `allow_failure(quote("foo"))` — maybe I should make the context have a list or set of annotations?

Reply:
That makes sense to me - it seems that you could even change the key name to `annotations` without any backwards compatible breakage since the flow is re-parsed with each run

Reply:
I'm going to wait for this to be an actual issue so we can keep this simple and get it out there